assembly descriptors should be optional

The target image may just want to set a few env entries or a different command
 
```
[ERROR] Failed to execute goal org.jolokia:docker-maven-plugin:0.10.4:build (build-docker-image) on project wildfly-camel-itests-openshift-domain: Error reading assembly: No assembly descriptors found. -> [Help 1]
```

--- a/README.md
+++ b/README.md
@@ -5,7 +5,7 @@
 [![Flattr](http://api.flattr.com/button/flattr-badge-large.png)](http://flattr.com/thing/73919/Jolokia-JMX-on-Capsaicin)
 
 This is a Maven plugin for managing Docker images and containers for your builds.
-The current version ist **0.11.0** and works with Maven 3.2.1 and Docker 1.3.0 or later.
+The current version ist **0.11.1** and works with Maven 3.2.1 and Docker 1.3.0 or later.
 
 The current Docker API version used is `v1.15` (which is the minimal required API version).
 See the **[User Manual](doc/manual.md)** for details on how to override this value for new

--- a/doc/changelog.md
+++ b/doc/changelog.md
@@ -1,5 +1,8 @@
 # ChangeLog
 
+* **0.11.2**
+  - allow the `assembly` configuraiton to be optional (#66)
+
 * **0.11.1**
   - Add support for binding UDP ports (#83)
   - "Entrypoint" supports now arguments (#84)

--- a/doc/manual.md
+++ b/doc/manual.md
@@ -284,6 +284,9 @@ Here's an example:
 * **ignorePermissions** indicates if existing file permissions should be ignored
   when creating the assembly archive. This value is `false` by default.
 
+In the event you do not need to include any artifacts with the image, you may
+safely omit this element from the configuration. 
+
 ##### Docker Assembly
 
 With using the `descriptor` or `descriptorRef` option

--- a/pom.xml
+++ b/pom.xml
@@ -6,7 +6,7 @@
 
   <groupId>org.jolokia</groupId>
   <artifactId>docker-maven-plugin</artifactId>
-  <version>0.11.1-SNAPSHOT</version>
+  <version>0.11.1</version>
   <packaging>maven-plugin</packaging>
 
   <name>maven-docker-plugin</name>
@@ -264,7 +264,7 @@
   <scm>
     <connection>scm:git:git://github.com/rhuss/docker-maven-plugin.git</connection>
     <developerConnection>scm:git:ssh://git@github.com/rhuss/docker-maven-plugin.git</developerConnection>
-    <tag>HEAD</tag>
+    <tag>v0.11.1</tag>
     <url>git://github.com/rhuss/docker-maven-plugin.git</url>
   </scm>
 

--- a/pom.xml
+++ b/pom.xml
@@ -6,7 +6,7 @@
 
   <groupId>org.jolokia</groupId>
   <artifactId>docker-maven-plugin</artifactId>
-  <version>0.11.1</version>
+  <version>0.11.2-SNAPSHOT</version>
   <packaging>maven-plugin</packaging>
 
   <name>maven-docker-plugin</name>
@@ -264,7 +264,7 @@
   <scm>
     <connection>scm:git:git://github.com/rhuss/docker-maven-plugin.git</connection>
     <developerConnection>scm:git:ssh://git@github.com/rhuss/docker-maven-plugin.git</developerConnection>
-    <tag>v0.11.1</tag>
+    <tag>HEAD</tag>
     <url>git://github.com/rhuss/docker-maven-plugin.git</url>
   </scm>
 

--- a/samples/cargo-jolokia-demo/pom.xml
+++ b/samples/cargo-jolokia-demo/pom.xml
@@ -21,7 +21,7 @@
 
   <groupId>org.jolokia</groupId>
   <artifactId>docker-jolokia-demo</artifactId>
-  <version>0.11.0</version>
+  <version>0.11.1</version>
 
   <url>http://www.jolokia.org</url>
 

--- a/samples/data-jolokia-demo/pom.xml
+++ b/samples/data-jolokia-demo/pom.xml
@@ -21,7 +21,7 @@
 
   <groupId>org.jolokia</groupId>
   <artifactId>docker-jolokia-demo</artifactId>
-  <version>0.11.1-SNAPSHOT</version>
+  <version>0.11.1</version>
 
   <url>http://www.jolokia.org</url>
 

--- a/src/main/java/org/jolokia/docker/maven/assembly/DockerAssemblyConfigurationSource.java
+++ b/src/main/java/org/jolokia/docker/maven/assembly/DockerAssemblyConfigurationSource.java
@@ -43,6 +43,10 @@ public class DockerAssemblyConfigurationSource implements AssemblerConfiguration
 
     @Override
     public String[] getDescriptorReferences() {
+        if (assemblyConfig.isEmpty()) {
+            return new String[] { "empty" };
+        }
+        
         String descriptorRef = assemblyConfig.getDescriptorRef();
         return (descriptorRef != null) ? new String[] { descriptorRef } : null;
     }

--- a/src/main/java/org/jolokia/docker/maven/assembly/DockerAssemblyManager.java
+++ b/src/main/java/org/jolokia/docker/maven/assembly/DockerAssemblyManager.java
@@ -53,18 +53,20 @@ public class DockerAssemblyManager {
         File outputDir = source.getOutputDirectory();
 
         try {
-            // Create assembly in output directory (specified in source)
             createArchiveFromAssembly(source);
             
             String dockerFileDir = assemblyConfig.getDockerFileDir();
             if (dockerFileDir != null) {
                 // Use specified docker directory which must include a Dockerfile.
                 return createTarball(source, validateDockerDir(params, dockerFileDir));
-            } else {
-                // Create custom docker file in output dir
-                createDockerFile(buildConfig, assemblyConfig, outputDir);
-                return createTarball(source,null);
-            }
+            } 
+
+            // Create custom docker file in output dir
+            DockerFileBuilder builder = createDockerFileBuilder(buildConfig, assemblyConfig);
+            builder.write(outputDir);
+            
+            return createTarball(source, null);
+            
         } catch (IOException e) {
             throw new MojoExecutionException(String.format("Cannot create Dockerfile in %s", outputDir), e);
         }
@@ -97,17 +99,23 @@ public class DockerAssemblyManager {
             throw new MojoExecutionException("Cannot create archive " + archive, e);
         }
     }
-
-    private void createDockerFile(BuildImageConfiguration buildConfig, AssemblyConfiguration assemblyConfig, File outputDir) 
-        throws IOException {
+    
+    // visible for testing
+    DockerFileBuilder createDockerFileBuilder(BuildImageConfiguration buildConfig, AssemblyConfiguration assemblyConfig) {
         DockerFileBuilder builder =
-                new DockerFileBuilder().basedir(assemblyConfig.getBasedir())
-                        .add("maven", "")
+                new DockerFileBuilder()
                         .env(buildConfig.getEnv())
-                        .exportBasedir(assemblyConfig.exportBasedir())
                         .expose(buildConfig.getPorts())
                         .volumes(buildConfig.getVolumes());
-        
+
+        if (assemblyConfig.isEmpty()) {
+            builder.emptyArchive();
+        } else {
+            builder.add("maven", "")
+                    .basedir(assemblyConfig.getBasedir())
+                    .exportBasedir(assemblyConfig.exportBasedir());
+        }
+
         if (buildConfig.getFrom() != null) {
             builder.baseImage(buildConfig.getFrom());
             builder.command((String[]) null); // Use command from base image (gets overwritten below if explicitly set)
@@ -118,8 +126,8 @@ public class DockerAssemblyManager {
         if (buildConfig.getCommand() != null) {
             builder.command(EnvUtil.splitWOnSpaceWithEscape(buildConfig.getCommand()));
         }
-        
-        builder.write(outputDir);
+
+        return builder;
     }
 
     private void createArchiveFromAssembly(DockerAssemblyConfigurationSource source) throws MojoExecutionException {

--- a/src/main/java/org/jolokia/docker/maven/assembly/DockerFileBuilder.java
+++ b/src/main/java/org/jolokia/docker/maven/assembly/DockerFileBuilder.java
@@ -29,6 +29,8 @@ public class DockerFileBuilder {
 
     private boolean exportBasedir = true;
     
+    private boolean emptyArchive;
+    
     // List of files to add. Source and destination follow except that destination
     // in interpreted as a relative path to the exportDir
     // See also http://docs.docker.io/reference/builder/#add
@@ -61,7 +63,7 @@ public class DockerFileBuilder {
      * @throws IllegalArgumentException if no src/dest entries have been added
      */
     public String content() throws IllegalArgumentException {
-        if (addEntries.size() == 0) {
+        if (!emptyArchive && addEntries.size() == 0) {
             throw new IllegalArgumentException("No entries added");
         }
 
@@ -188,6 +190,13 @@ public class DockerFileBuilder {
         return this;
     }
 
+    public DockerFileBuilder emptyArchive() {
+        this.emptyArchive = true;
+        this.exportBasedir = false;
+
+        return this;
+    }
+    
     public DockerFileBuilder exportBasedir(boolean exportBasedir) {
         this.exportBasedir = exportBasedir;
         return this;

--- a/src/main/java/org/jolokia/docker/maven/config/AssemblyConfiguration.java
+++ b/src/main/java/org/jolokia/docker/maven/config/AssemblyConfiguration.java
@@ -3,8 +3,10 @@ package org.jolokia.docker.maven.config;
 
 public class AssemblyConfiguration {
 
+    public static final AssemblyConfiguration EMPTY = new AssemblyConfiguration(true);
+    
     private static final String DEFAULT_BASE_DIR = "/maven";
-
+    
     /**
      * @parameter
      */
@@ -39,10 +41,22 @@ public class AssemblyConfiguration {
      * @parameter
      */
     private String user;
+    
+    private Boolean empty;
+    
+    public AssemblyConfiguration() {  
+        this(false);
+    }
 
-    public AssemblyConfiguration() { }
-
+    private AssemblyConfiguration(boolean empty) {
+        this.empty = empty;
+    }
+    
     public Boolean exportBasedir() {
+        if (empty) {
+            return false;
+        }
+        
         return exportBasedir != null ? exportBasedir : Boolean.TRUE;
     }
 
@@ -50,7 +64,7 @@ public class AssemblyConfiguration {
         return basedir != null ? basedir : DEFAULT_BASE_DIR;
     }
 
-    public String getDescriptor() {
+    public String getDescriptor() {        
         return descriptor;
     }
 
@@ -66,13 +80,17 @@ public class AssemblyConfiguration {
         return user;
     }
 
+    public Boolean isEmpty() {
+        return empty;
+    }
+    
     public Boolean isIgnorePermissions() {
         return (ignorePermissions != null) ? ignorePermissions : Boolean.FALSE;
     }
-            
+    
     public static class Builder {
 
-        private final AssemblyConfiguration config = new AssemblyConfiguration();
+        private final AssemblyConfiguration config = new AssemblyConfiguration(true);
 
         public Builder basedir(String baseDir) {
             config.basedir = baseDir;
@@ -85,11 +103,15 @@ public class AssemblyConfiguration {
 
         public Builder descriptor(String descriptor) {
             config.descriptor = descriptor;
+            setEmpty(descriptor);
+            
             return this;
         }
 
         public Builder descriptorRef(String descriptorRef) {
             config.descriptorRef = descriptorRef;
+            setEmpty(descriptorRef);
+            
             return this;
         }
 
@@ -97,7 +119,7 @@ public class AssemblyConfiguration {
             config.dockerFileDir = dockerFileDir;
             return this;
         }
-
+        
         public Builder exportBasedir(Boolean export) {
             config.exportBasedir = export;
             return this;
@@ -111,6 +133,12 @@ public class AssemblyConfiguration {
         public Builder user(String user) {
             config.user = user;
             return this;
+        }
+        
+        private void setEmpty(String descriptorOrRef) {
+            if (descriptorOrRef != null) {
+                config.empty = false;
+            }
         }
     }
 }

--- a/src/main/java/org/jolokia/docker/maven/config/BuildImageConfiguration.java
+++ b/src/main/java/org/jolokia/docker/maven/config/BuildImageConfiguration.java
@@ -57,6 +57,10 @@ public class BuildImageConfiguration {
     }
 
     public AssemblyConfiguration getAssemblyConfiguration() {
+        if (assembly == null) {
+            return AssemblyConfiguration.EMPTY;
+        }
+
         return assembly;
     }
 

--- a/src/main/resources/assemblies/empty.xml
+++ b/src/main/resources/assemblies/empty.xml
@@ -1,0 +1,16 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<assembly
+  xmlns="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/${mdoVersion}"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/${mdoVersion} http://maven.apache.org/xsd/assembly-${mdoVersion}.xsd">
+  <id>project</id>
+  <fileSets>
+    <fileSet>
+      <directory>./</directory>
+      <outputDirectory>/</outputDirectory>
+      <excludes>
+        <exclude>*/**</exclude>
+      </excludes>
+    </fileSet>
+  </fileSets>
+</assembly>

--- a/src/test/java/org/jolokia/docker/maven/assembly/DockerAssemblyManagerTest.java
+++ b/src/test/java/org/jolokia/docker/maven/assembly/DockerAssemblyManagerTest.java
@@ -1,0 +1,30 @@
+package org.jolokia.docker.maven.assembly;
+
+import static org.junit.Assert.assertFalse;
+
+import org.jolokia.docker.maven.config.AssemblyConfiguration;
+import org.jolokia.docker.maven.config.BuildImageConfiguration;
+import org.junit.Before;
+import org.junit.Test;
+
+public class DockerAssemblyManagerTest {
+
+    private DockerAssemblyManager assemblyManager;
+
+    @Before
+    public void setup() {
+        assemblyManager = new DockerAssemblyManager();
+    }
+
+    @Test
+    public void testNoAssembly() {
+        BuildImageConfiguration buildConfig = new BuildImageConfiguration();
+        AssemblyConfiguration assemblyConfig = buildConfig.getAssemblyConfiguration();
+
+        DockerFileBuilder builder = assemblyManager.createDockerFileBuilder(buildConfig, assemblyConfig);
+        String content = builder.content();
+
+        assertFalse(content.contains("COPY"));
+        assertFalse(content.contains("VOLUME"));
+    }
+}

--- a/src/test/java/org/jolokia/docker/maven/config/handler/PropertyConfigHandlerTest.java
+++ b/src/test/java/org/jolokia/docker/maven/config/handler/PropertyConfigHandlerTest.java
@@ -108,6 +108,17 @@ public class PropertyConfigHandlerTest {
     }
     
     @Test
+    public void testNoAssembly() throws Exception {
+        Properties props = props(new String[] { k(NAME), "image" });
+        List<ImageConfiguration> configs = configHandler.resolve(imageConfiguration, props);
+        assertEquals(1, configs.size());
+
+        AssemblyConfiguration config = configs.get(0).getBuildConfiguration().getAssemblyConfiguration();
+        assertTrue(config.isEmpty());
+        assertFalse(config.exportBasedir());
+    }
+    
+    @Test
     public void testResolve() {
         Map<String, String> external = new HashMap<>();
         external.put("type", "props");

